### ISSUE TITLE
feature: Dynamic pages

### ISSUE THREAD

**Issue by Mange** [Mange/tydra#7]
It could be useful to have a page that could contain dynamic information on it.

Some examples:
- In a page for media control
  - Show the currently playing media in the header.
  - Switch between `[p] Pause` and `[p] Play` depending on player state.

Since the YAML file could be generated before starting tydra (`tydra <(generate_actions)`) you could get the basics up and running fairly quickly, but that also limits everything a bit because it cannot be refreshed until `tydra` is restarted.

One way of handling this is to allow a page to be defined as just being a command that should render the page definition. This command would then run before the page is displayed, allowing it to have the status up-to-date.

Considering the complexity of building a system for this it might not be worth it. However, there might be a "simple" escape hatch one could use to let users build this system themselves if they so wish:

A `--restart-on-return` flag could be provided, which would reload `tydra` when returning. Then clients would write their wrapper scripts that perform an action and then regenerates the config before tydra is finally restarted.

**Comment by Mange** [comment]
Original hydra has support for adding dynamic values to text using %\`:

> To insert a dynamic Elisp variable, use %\` followed by the variable. Each time the variable changes due to a head, the docstring will be updated. format-style width specifiers can be used.
>
> To insert a dynamic Elisp expression, use e.g. %(length (dired-get-marked-files)). If a head will change the amount of marked files, for example, it will be appropriately updated.
>
> If the result of the Elisp expression is a string and you don't want to quote it, use this form: %s(shell-command-to-string "du -hs").

Maybe doing this in a markup fashion is also a possible way forward? Markup is already planned for changing colors, etc.

**Comment by edrex** [comment]
What about
```
        - shortcut: s
          title: Submenu
          command: tydra <(cat submenu.yaml)
          return: true
        - shortcut: r
          title: Reload
          mode: exec
          command: tydra actions.yaml
```
?

I like really like the simplicity and orthogonality of your design of tydra. I know it largely comes from hydra, but you adapted it to the terminal context really well.

Also, I've been shell scripting for a couple decades and the `<(cmd)` syntax for getting a file descriptor/named pipe didn't stick until now, so thanks for that.

**Comment by Mange** [comment]
> What about
> 
> ```yaml
>           command: tydra <(cat submenu.yaml)
> ```
> 
> ?

Good idea. It's probably an even better way to do it for a KISS solution.

> I like really like the simplicity and orthogonality of your design of tydra. I know it largely comes from hydra, but you adapted it to the terminal context really well.

Thank you!

Sadly I've stopped using it completely myself at this point, so no new feature is likely to be added from here on out.

> Also, I've been shell scripting for a couple decades and the `<(cmd)` syntax for getting a file descriptor/named pipe didn't stick until now, so thanks for that.

It's super great for things like `diff -u <(sort expected.txt) <(curl … | sort)`. :smile:

**Comment by edrex** [comment]
> Sadly I've stopped using it completely myself at this point, so no new feature is likely to be added from here on out.

TBH I was doing one last search for prior art before starting my own hydra-in-terminal implementation (was thinking to use Zig) when I found tydra. I really like the design and implementation, so i think I'll be using it ongoing, and patching as needed.

**Comment by Mange** [comment]
Let me know if you'd be interested in becoming an official maintainer. You could always do an actual fork as well, of course. :-)

**Comment by edrex** [comment]
There are two use-cases I'm primarily interested in:

 - using tydra to prototype a "conversational, contextual shell" (to be coupled with a context/activity tracking service, which is a big research topic)
 - to provide a "spacemacs-like" shell experience on top of nix flakes
